README
Overview
========

DiskGraph is a Python utility that creates a graph of how disks, partitions, etc.
relate to each other in a Linux system. The following entities are supported:

* Disks and partitions (hd/sd, read from /proc/partitions)
* Raid (read from /proc/mdstat)
* LVM (read from execution of pvs and lvs)

Because the pvs and lvs commands must be run as root, this utility must as well.

Disclaimer
==========

See the LICENSE file. The utility must be run as root; read the source code
carefully before running it if you're uncertain as to what it does.

Prerequisites
=============

pydot (python-pydot)

Usage
=====

To run it:

sudo diskgraph/dgmain.py diskgraph.png

Since the only way to watc the output is to open the PNG file, a suggested usage
on a server is:

sudo diskgraph/dgmain.py /var/www/diskgraph.png



CHANGELOG
v1.2 (2012-01-02)
* Increased robustness wrt missing commands or files (issue #1)
* Added inclusion of swap areas
* Automatic to manual free space expansion to avoid irrelevant free space
* Some improved node labels

v1.1 (2011-12-28)
* Coloring of graph nodes
* Nodes for free space >= 5 MB included in the graph
* Mounted file systems included in the graph

v1.0 (2011-12-27)
* First version



test/test_sgraph.py
import unittest
from diskgraph.sgraph import *

class TestSimpleGraph(unittest.TestCase):

    def test_that_graph_is_created_with_head_finder_and_root(self):
        root = 1
        headfinder = lambda x: []
        self.assertIsInstance(SimpleGraph(headfinder, root), SimpleGraph)

    def test_that_graph_exposes_root(self):
        root = 1
        g = SimpleGraph(lambda x: [], root)
        self.assertEquals(root, g.root)

    def test_that_graph_with_single_vertex_can_be_built(self):
        root = 1
        headfinder = lambda x: []
        graph = SimpleGraph(headfinder, root)
        self.assertEqual(1, graph.order)

    def test_that_graph_with_two_vertices_can_be_built(self):
        root = 1
        headfinder = lambda x: [2] if x == root else []
        graph = SimpleGraph(headfinder, root)
        self.assertEqual(2, graph.order)

    def test_that_graph_with_loop_can_be_built(self):
        root = 1
        headfinder = lambda x: [2] if x == root else [root]
        graph = SimpleGraph(headfinder, root)
        self.assertEqual(2, graph.order)

    def test_that_headsFor_handles_loop_case(self):
        root = 1
        headfinder = lambda x: [2] if x == root else [root]
        graph = SimpleGraph(headfinder, root)
        self.assertEqual([root], graph.headsFor(2))

    def test_that_heads_for_vertex_can_be_queried(self):
        root = 1
        headfinder = lambda x: [2] if x == root else []
        graph = SimpleGraph(headfinder, root)
        heads = graph.headsFor(1)
        self.assertEqual([2], heads)

    def test_that_tails_for_vertex_can_be_queried(self):
        root = 1
        headfinder = lambda x: [2] if x == root else []
        graph = SimpleGraph(headfinder, root)
        heads = graph.tailsFor(2)
        self.assertEqual([1], heads)

    def test_that_vertices_can_be_visited(self):
        root = 1
        headfinder = lambda x: [2] if x == root else []
        graph = SimpleGraph(headfinder, root)
        visited = list(graph.visit(root))
        self.assertEqual([1, 2], visited)

    def test_that_loops_are_handled_when_visiting(self):
        root = 1
        headfinder = lambda x: [2] if x == root else [root]
        graph = SimpleGraph(headfinder, root)
        visited = list(graph.visit(root))
        self.assertEqual([1, 2], visited)

    def test_that_heads_are_visited_ltr(self):
        root = 1
        headfinder = lambda x: [2, 3] if x == root else []
        graph = SimpleGraph(headfinder, root)
        visited = list(graph.visit(root))
        self.assertEqual([root, 2, 3], visited)

    def test_that_edges_can_be_visited_for_simple_graph(self):
        root = 1
        headfinder = lambda x: [2] if x == root else []
        graph = SimpleGraph(headfinder, root)
        visited = list(graph.visitEdges(root))
        self.assertEqual([(root, 2)], visited)

    def test_that_edges_can_be_visited_for_graph_with_loop(self):
        root = 1
        headfinder = lambda x: [2] if x == root else [root]
        graph = SimpleGraph(headfinder, root)
        visited = list(graph.visitEdges(root))
        self.assertEqual([(root, 2), (2, root)], visited)

    def test_that_head_can_be_added_for_vertex(self):
        root = 1
        headfinder = lambda x: [2] if x == root else []
        graph = SimpleGraph(headfinder, root)
        graph.addHead(root, 3)
        self.assertTrue(3 in graph.headsFor(root))




test/test_sysinfo.py
import re
import unittest
from diskgraph.sysinfo import *
from mock import patch, MagicMock, Mock
from cStringIO import StringIO
from diskgraph.check import Checker
from subprocess import CalledProcessError

def checker_mock(b):
    c = Mock(spec=Checker)
    c.has_mdstat.return_value = b
    c.has_df_command.return_value = b
    c.has_partitions.return_value = b
    c.has_lvm_commands.return_value = b
    c.has_swaps.return_value = b
    return c

def splitkeepsep(s, sep):
    return reduce(lambda acc, i: acc[:-1] + [acc[-1] + i] if i == sep else acc + [i], re.split("(%s)" % re.escape(sep), s), [])

def confopenmock(mock, text):
    mock.return_value = MagicMock(spec=file)
    handle = mock.return_value.__enter__.return_value
    handle.__iter__.return_value = iter(splitkeepsep(text, "\n"))

class TestSysInfoPartitionGeneration(unittest.TestCase):
    @patch("diskgraph.sysinfo.checker", checker_mock(True))
    @patch("diskgraph.sysinfo.open", create=True)
    def setUp(self, open_mock):
        text = "major minor  #blocks  name\n\n   8        0  244198584 sda\n"
        confopenmock(open_mock, text)
        self.pp = Partition.generate()

    def test_that_one_partition_is_found(self):
        self.assertEqual(1, len(self.pp))

    def test_that_partition_is_of_right_type(self):
        part = self.pp[0]
        self.assertIsInstance(part, Partition)

    def test_that_partition_contains_kernel_major_minor(self):
        part = self.pp[0]
        self.assertEqual((8, 0), part.kernel_major_minor)

    def test_that_partition_contains_name(self):
        part = self.pp[0]
        self.assertEqual("sda", part.name)

    def test_that_partition_contains_size(self):
        part = self.pp[0]
        self.assertEqual(250059350016, part.byte_size)

class TestSysInfoRaidArrayGeneration(unittest.TestCase):
    @patch("diskgraph.sysinfo.checker", checker_mock(True))
    @patch("diskgraph.sysinfo.open", create=True)
    def setUp(self, open_mock):
        text = ("Personalities : [raid1] [raid6] [raid5] [raid4] [linear] [multipath] [raid0] [raid10]\n"
                "md1 : active raid1 sdf1[1] sdi1[0]\n"
                "      244195904 blocks [2/2] [UU]\n")
        confopenmock(open_mock, text)
        self.md = RaidArray.generate()

    def test_that_one_array_is_found(self):
        self.assertEqual(1, len(self.md))

    def test_that_array_is_of_right_type(self):
        arr = self.md[0]
        self.assertIsInstance(arr, RaidArray)

    def test_that_array_contains_name(self):
        arr = self.md[0]
        self.assertEqual("md1", arr.name)

    def test_that_array_contains_partition_names(self):
        arr = self.md[0]
        self.assertListEqual(["sdf1", "sdi1"], arr.partition_names)

    def test_that_array_contains_size(self):
        arr = self.md[0]
        self.assertEqual(250056605696, arr.byte_size)

class TestSysInfoSwapAreaGeneration(unittest.TestCase):
    @patch("diskgraph.sysinfo.checker", checker_mock(True))
    @patch("diskgraph.sysinfo.open", create=True)
    def setUp(self, open_mock):
        text = ("Filename                                Type            Size    Used    Priority\n"
                "/dev/sda2                               partition       497660  0       -1\n")
        confopenmock(open_mock, text)
        self.swaps = SwapArea.generate()

    def test_that_one_swap_area_is_found(self):
        self.assertEqual(1, len(self.swaps))

    def test_that_swap_is_of_right_type(self):
        swap = self.swaps[0]
        self.assertIsInstance(swap, SwapArea)

    def test_that_swap_area_contains_name(self):
        swap = self.swaps[0]
        self.assertEqual("sda2", swap.name)

    def test_that_swap_area_contains_size(self):
        swap = self.swaps[0]
        self.assertEqual(509603840, swap.byte_size)

class TestSysInfoMountedFileSystemGeneration(unittest.TestCase):
    @patch("diskgraph.sysinfo.checker", checker_mock(True))
    @patch("subprocess.check_output")
    def setUp(self, exec_mock):
        exec_mock.return_value = ("Filesystem           1B-blocks      Used Available Use% Mounted on\n"
                                  "/dev/sdk2            3897212928 2526269440 1212547072  68% /boot\n")
        self.mounts = MountedFileSystem.generate()

    def test_that_one_mounted_fs_is_found(self):
        self.assertEqual(1, len(self.mounts))

    def test_that_mounted_fs_is_of_right_type(self):
        mfs = self.mounts[0]
        self.assertIsInstance(mfs, MountedFileSystem)

    def test_that_mounted_fs_contains_name(self):
        mfs = self.mounts[0]
        self.assertEqual("/boot", mfs.name)

    def test_that_mounted_fs_contains_size(self):
        mfs = self.mounts[0]
        self.assertEqual(3897212928, mfs.byte_size)

class TestSwapArea(unittest.TestCase):
    def test_that_swap_area_is_child_of_partition(self):
        p = Partition("8 2 497660 sda2".split(" "))
        sa = SwapArea("/dev/sda2 partition 497660 0 -1".split(" "))
        self.assertTrue(sa.is_child_of(p))

class TestMountedFileSystem(unittest.TestCase):
    def test_that_mounted_fs_is_not_child_of_wrong_partition(self):
        p = Partition("8 1 1000 sda1".split(" "))
        mfs = MountedFileSystem("/dev/sda2 3897212928 2526269440 1212547072 68% /boot".split(" "))
        self.assertFalse(mfs.is_child_of(p))

    def test_that_mounted_fs_is_child_of_partition(self):
        p = Partition("8 1 1000 sda1".split(" "))
        mfs = MountedFileSystem("/dev/sda1 3897212928 2526269440 1212547072 68% /boot".split(" "))
        self.assertTrue(mfs.is_child_of(p))

    def test_that_mounted_fs_is_child_of_raid_array(self):
        r = RaidArray(("md1 /dev/sda1".split(" "), 1000))
        mfs = MountedFileSystem("/dev/md1 3897212928 2526269440 1212547072 68% /boot".split(" "))
        self.assertTrue(mfs.is_child_of(r))

    def test_that_mounted_fs_is_child_of_lvm_lv_given_vg_matches(self):
        lv = LvmLogicalVolume("small test 1000".split(" "))
        mfs = MountedFileSystem("/dev/mapper/test-small 3897212928 2526269440 1212547072 68% /boot".split(" "))
        self.assertTrue(mfs.is_child_of(lv))

    def test_that_mounted_fs_is_not_child_of_lvm_lv_if_lv_mismatch(self):
        lv = LvmLogicalVolume("big test 1000".split(" "))
        mfs = MountedFileSystem("/dev/mapper/test-small 3897212928 2526269440 1212547072 68% /boot".split(" "))
        self.assertFalse(mfs.is_child_of(lv))

    def test_that_mounted_fs_is_not_child_of_lvm_lv_if_vg_mismatch(self):
        lv = LvmLogicalVolume("small blob 1000".split(" "))
        mfs = MountedFileSystem("/dev/mapper/test-small 3897212928 2526269440 1212547072 68% /boot".split(" "))
        self.assertFalse(mfs.is_child_of(lv))

    def test_that_mounted_fs_is_not_child_of_root(self):
        r = Root()
        mfs = MountedFileSystem("/dev/mapper/test-small 3897212928 2526269440 1212547072 68% /boot".split(" "))
        self.assertFalse(mfs.is_child_of(r))

class TestSysInfoLvmPhysicalVolumeGeneration(unittest.TestCase):
    @patch("diskgraph.sysinfo.checker", checker_mock(True))
    @patch("subprocess.check_output")
    def setUp(self, exec_mock):
        exec_mock.return_value = "  /dev/md0   1500310929408\n"
        self.pvs = LvmPhysicalVolume.generate()

    def test_that_one_lvm_physical_volume_is_found(self):
        self.assertEqual(1, len(self.pvs))

    def test_that_lvm_physical_volume_is_of_right_type(self):
        pv = self.pvs[0]
        self.assertIsInstance(pv, LvmPhysicalVolume)

    def test_that_lvm_physical_volume_contains_name(self):
        pv = self.pvs[0]
        self.assertEqual("md0", pv.name)

    def test_that_lvm_physical_volume_contains_size(self):
        pv = self.pvs[0]
        self.assertEqual(1500310929408, pv.byte_size)

class TestSysInfoLvmVolumeGroupGeneration(unittest.TestCase):
    @patch("diskgraph.sysinfo.checker", checker_mock(True))
    @patch("subprocess.check_output")
    def setUp(self, exec_mock):
        exec_mock.return_value = ("  backup  700146778112 /dev/md1 0\n"
                                  "  backup  700146778112 /dev/md2 0\n"
                                  "  small    50008686592 /dev/sdd2 0\n")
        self.vgs = LvmVolumeGroup.generate()

    def test_that_two_lvm_volume_groups_are_found(self):
        self.assertEqual(2, len(self.vgs))

    def test_that_lvm_volume_group_is_of_right_type(self):
        vg = self.vgs[0]
        self.assertIsInstance(vg, LvmVolumeGroup)

    def test_that_lvm_volume_group_contains_name(self):
        vg = self.vgs[0]
        self.assertEqual("backup", vg.name)

    def test_that_lvm_volume_group_contains_size(self):
        vg = self.vgs[0]
        self.assertEqual(700146778112, vg.byte_size)

    def test_that_lvm_volume_group_contains_physical_volume_names(self):
        vg = self.vgs[0]
        self.assertListEqual(["md1", "md2"], vg.pv_names)

class TestSysInfoLvmLogicalVolumeGeneration(unittest.TestCase):
    @patch("diskgraph.sysinfo.checker", checker_mock(True))
    @patch("subprocess.check_output")
    def setUp(self, exec_mock):
        exec_mock.return_value = "  homes   backup   21474836480\n"
        self.lvs = LvmLogicalVolume.generate()

    def test_that_one_lvm_logical_volume_is_found(self):
        self.assertEqual(1, len(self.lvs))

    def test_that_lvm_logical_volume_is_of_right_type(self):
        lv = self.lvs[0]
        self.assertIsInstance(lv, LvmLogicalVolume)

    def test_that_lvm_logical_volume_contains_name(self):
        lv = self.lvs[0]
        self.assertEqual("homes", lv.name)

    def test_that_lvm_logical_volume_contains_size(self):
        lv = self.lvs[0]
        self.assertEqual(21474836480, lv.byte_size)

    def test_that_lvm_logical_volume_contains_volume_group_name(self):
        lv = self.lvs[0]
        self.assertEqual("backup", lv.vg_name)

class TestPartition(unittest.TestCase):
    def test_that_hd_without_number_is_disk(self):
        p = Partition("3 0 1000 hda".split(" "))
        self.assertTrue(p.is_disk())

    def test_that_sd_without_number_is_disk(self):
        p = Partition("8 0 1000 sda".split(" "))
        self.assertTrue(p.is_disk())

    def test_that_hd_with_number_is_not_disk(self):
        p = Partition("3 1 1000 hda1".split(" "))
        self.assertFalse(p.is_disk())

    def test_that_partition_is_part_for_disk(self):
        d = Partition("8 0 1000 sda".split(" "))
        p = Partition("8 1 1000 sda1".split(" "))
        self.assertTrue(p.is_partition_for(d))

    def test_that_partition_is_not_part_for_disk(self):
        d = Partition("8 0 1000 sda".split(" "))
        p = Partition("8 17 1000 sdb1".split(" "))
        self.assertFalse(p.is_partition_for(d))

    def test_that_partition_is_child_of_root_if_disk(self):
        r = Root()
        d = Partition("8 0 1000 sda".split(" "))
        self.assertTrue(d.is_child_of(r))

    def test_that_partition_is_not_child_of_root_if_partition(self):
        r = Root()
        p = Partition("8 1 1000 sda1".split(" "))
        self.assertFalse(p.is_child_of(r))

    def test_that_partition_is_child_of_disk(self):
        d = Partition("8 0 1000 sda".split(" "))
        p = Partition("8 1 1000 sda1".split(" "))
        self.assertTrue(p.is_child_of(d))

class TestLvmPhysicalVolume(unittest.TestCase):
    def test_that_pv_is_child_of_disk_if_name_matches(self):
        d = Partition("8 0 1000 sda".split(" "))
        pv = LvmPhysicalVolume("/dev/sda 1000".split(" "))
        self.assertTrue(pv.is_child_of(d))

    def test_that_pv_is_not_child_of_disk_if_name_doesnt_match(self):
        d = Partition("8 0 1000 sda".split(" "))
        pv = LvmPhysicalVolume("/dev/sdb 1000".split(" "))
        self.assertFalse(pv.is_child_of(d))

    def test_that_pv_is_child_of_partition(self):
        p = Partition("8 1 1000 sda1".split(" "))
        pv = LvmPhysicalVolume("/dev/sda1 1000".split(" "))
        self.assertTrue(pv.is_child_of(p))

    def test_that_pv_is_child_of_raid_array(self):
        r = RaidArray(("md1 /dev/sda1".split(" "), 1000))
        pv = LvmPhysicalVolume("/dev/md1 1000".split(" "))
        self.assertTrue(pv.is_child_of(r))

    def test_that_pv_is_not_child_of_root(self):
        r = Root()
        pv = LvmPhysicalVolume("/dev/sda 1000".split(" "))
        self.assertFalse(pv.is_child_of(r))

class LvmVolumeGroupTest(unittest.TestCase):
    def test_that_vg_is_child_of_pv_if_pv_among_names(self):
        pv = LvmPhysicalVolume("/dev/sda 1000".split(" "))
        vg = LvmVolumeGroup(["test", "1000", ["/dev/sda"], "0"])
        self.assertTrue(vg.is_child_of(pv))

    def test_that_vg_is_not_child_of_pv_if_pv_not_among_names(self):
        pv = LvmPhysicalVolume("/dev/sda 1000".split(" "))
        vg = LvmVolumeGroup(["test", "1000", ["/dev/sdb"], "0"])
        self.assertFalse(vg.is_child_of(pv))

    def test_that_vg_is_not_child_of_root(self):
        r = Root()
        vg = LvmVolumeGroup(["test", "1000", ["/dev/sdb"], "0"])
        self.assertFalse(vg.is_child_of(r))

class LvmLogicalVolumeTest(unittest.TestCase):
    def test_that_lv_is_child_of_vg_with_correct_name(self):
        vg = LvmVolumeGroup(["test", "1000", ["/dev/sdb"], "0"])
        lv = LvmLogicalVolume("small test 1000".split(" "))
        self.assertTrue(lv.is_child_of(vg))

    def test_that_lv_is_not_child_of_vg_with_wrong_name(self):
        vg = LvmVolumeGroup(["blob", "1000", ["/dev/sdb"], "0"])
        lv = LvmLogicalVolume("small test 1000".split(" "))
        self.assertFalse(lv.is_child_of(vg))

    def test_that_lv_is_not_child_of_root(self):
        r = Root()
        lv = LvmLogicalVolume("small test 1000".split(" "))
        self.assertFalse(lv.is_child_of(r))

class RaidArrayTest(unittest.TestCase):
    def test_that_array_is_child_of_disk_among_names(self):
        d = Partition("8 0 1000 /dev/sda".split(" "))
        r = RaidArray(("md1 /dev/sda".split(" "), 1000))
        self.assertTrue(r.is_child_of(d))

    def test_that_array_is_not_child_of_disk_not_among_names(self):
        d = Partition("8 16 1000 /dev/sdb".split(" "))
        r = RaidArray(("md1 /dev/sda".split(" "), 1000))
        self.assertFalse(r.is_child_of(d))

    def test_that_array_is_child_of_partition_among_names(self):
        p = Partition("8 1 1000 /dev/sda1".split(" "))
        r = RaidArray(("md1 /dev/sda1".split(" "), 1000))
        self.assertTrue(r.is_child_of(p))

@patch("diskgraph.sysinfo.checker", checker_mock(False))
class MissingFileOrCommandTest(unittest.TestCase):
    @patch("subprocess.check_output")
    def test_that_no_lvm_physical_volumes_are_found_if_lvm_commands_dont_exist(self, co_mock):
        co_mock.side_effect = CalledProcessError(1, "pvs")
        self.assertEqual(0, len(LvmPhysicalVolume.generate()))

    @patch("subprocess.check_output")
    def test_that_no_lvm_logical_volums_are_found_if_lvm_commands_dont_exist(self, co_mock):
        co_mock.side_effect = CalledProcessError(1, "lvs")
        self.assertEqual(0, len(LvmLogicalVolume.generate()))

    @patch("subprocess.check_output")
    def test_that_no_lvm_volume_groups_are_found_if_lvm_commands_dont_exist(self, co_mock):
        co_mock.side_effect = CalledProcessError(1, "vgs")
        self.assertEqual(0, len(LvmVolumeGroup.generate()))

    @patch("diskgraph.sysinfo.open", create=True)
    def test_that_no_arrays_are_found_if_mdstat_file_doesnt_exist(self, open_mock):
        open_mock.side_effect = IOError
        self.assertEqual(0, len(RaidArray.generate()))

    @patch("subprocess.check_output")
    def test_that_no_mounted_fs_are_found_if_df_command_doesnt_exist(self, co_mock):
        co_mock.side_effect = CalledProcessError(1, "df")
        self.assertEqual(0, len(MountedFileSystem.generate()))
    
    @patch("diskgraph.sysinfo.open", create=True)
    def test_that_no_swap_areas_are_found_if_swaps_file_doesnt_exist(self, open_mock):
        open_mock.side_effect = IOError
        self.assertEqual(0, len(SwapArea.generate()))

class TestFreeSpace(unittest.TestCase):
    def test_that_disk_expands_free_space(self):
        parts = [Partition("8 0 204800 sda".split(" ")),
                 Partition("8 1 51200 sda1".split(" "))]
        expanded = parts[0].expand(parts)
        self.assertTrue(FreeSpace in [e.__class__ for e in expanded])

    def test_that_disk_calculates_free_space_correctly(self):
        parts = [Partition("8 0 204800 sda".split(" ")),
                 Partition("8 1 51200 sda1".split(" "))]
        expanded = parts[0].expand(parts)
        fs = [e for e in expanded if isinstance(e, FreeSpace)][0]
        self.assertEqual(157286400, fs.byte_size)

    def test_that_disk_doesnt_expand_small_free_space(self):
        parts = [Partition("8 0 204800 sda".split(" ")),
                 Partition("8 1 103424 sda1".split(" "))]
        expanded = parts[0].expand(parts)
        self.assertFalse(FreeSpace in [e.__class__ for e in expanded])

    def test_that_lvm_vg_expands_free_space(self):
        parts = [LvmVolumeGroup("test 209715200 /dev/sda1 157286400".split(" "))]
        expanded = parts[0].expand(parts)
        self.assertTrue(FreeSpace in [e.__class__ for e in expanded])

    def test_that_lvm_vg_calculates_free_space_correctly(self):
        parts = [LvmVolumeGroup("test 209715200 /dev/sda1 157286400".split(" "))]
        expanded = parts[0].expand(parts)
        fs = [e for e in expanded if isinstance(e, FreeSpace)][0]
        self.assertEqual(157286400, fs.byte_size)

    def test_that_lvm_vg_doesnt_expand_small_free_space(self):
        parts = [LvmVolumeGroup("test 209715200 /dev/sda1 103809024".split(" "))]
        expanded = parts[0].expand(parts)
        self.assertFalse(FreeSpace in [e.__class__ for e in expanded])

class TestToString(unittest.TestCase):
    def test_that_byte_size_is_optional(self):
        vg = LvmVolumeGroup("test 209715200 /dev/sda1 157286400".split(" "))
        del vg.byte_size # hack
        self.assertEqual("LvmVolumeGroup\ntest", str(vg))

    def test_that_type_name_and_size_are_included(self):
        vg = LvmVolumeGroup("test 209715200 /dev/sda1 157286400".split(" "))
        self.assertEqual("LvmVolumeGroup\ntest\n200.00MB", str(vg))

    def test_that_disk_has_custom_type(self):
        disk = Partition("8 0 204800 sda".split(" "))
        self.assertEqual("Disk\nsda\n200.00MB", str(disk))

    def test_that_free_space_has_custom_tostring(self):
        fs = FreeSpace(104857600)
        self.assertEqual("Free space\n100.00MB", str(fs))
    
    def test_that_root_has_custom_tostring(self):
        r = Root()
        self.assertEqual("Root", str(r))


diskgraph/check.py
# -*- coding: utf-8 -*-
"""Module for checking the availability of certain files and commands. Part of the
diskgraph utility.

Created and tested by Per Rovegård on a server running Ubuntu 11.04 with Python 2.7.1.

Distributed under the 3-Clause BSD license (http://opensource.org/licenses/BSD-3-Clause,
and LICENSE file).
"""

__author__ = "Per Rovegård"
__version__ = "1.2"
__license__ = "BSD-3-Clause"

import os.path
from subprocess import check_output, CalledProcessError

def file_exists(path):
    return os.path.exists(path)

def cmd_exists(cmd):
    try:
        check_output(["which", cmd])
        return True
    except CalledProcessError:
        return False

class Checker(object):
    def __init__(self):
        self._has_swaps = file_exists("/proc/swaps")
        self._has_partitions = file_exists("/proc/partitions")
        self._has_mdstat = file_exists("/proc/mdstat")
        self._has_lvm_commands = cmd_exists("pvs") and cmd_exists("vgs") and cmd_exists("lvs")
        self._has_df_command = cmd_exists("df")

    def has_partitions(self):
        return self._has_partitions

    def has_mdstat(self):
        return self._has_mdstat

    def has_lvm_commands(self):
        return self._has_lvm_commands

    def has_df_command(self):
        return self._has_df_command

    def has_swaps(self):
        return self._has_swaps

try:
    checker
except NameError:
    checker = Checker()



diskgraph/dgmain.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
"""Main script for diskgraph, a utility that creates a graph of disks, partitions, etc.
present on the current Linux server.

This script must be run as root, since the LVM commands require that.

Pydot (http://code.google.com/p/pydot) is used to create a PNG image of the graph.

Created and tested by Per Rovegård on a server running Ubuntu 11.04 with Python 2.7.1.

Distributed under the 3-Clause BSD license (http://opensource.org/licenses/BSD-3-Clause,
and LICENSE file).
"""

__author__ = "Per Rovegård"
__version__ = "1.2"
__license__ = "BSD-3-Clause"

import os, sys
from check import checker
from diskgraph import DiskGraph
from sysinfo import SysInfo

if not checker.has_partitions():
    sys.exit("The file /proc/partitions must exist.\n")

if not checker.has_df_command():
    print "No df command found - mounted file systems won't be included."

if not checker.has_mdstat():
    print "No /proc/mdstat file - software RAID arrays won't be included."

if not checker.has_swaps():
    print "No /proc/swaps file - swap areas won't be included."

if not checker.has_lvm_commands():
    print "No LVM commands founds - LVM entities won't be included."

# Currently, only the LVM commands require root privileges.
if checker.has_lvm_commands() and __name__ == "__main__":
    if os.geteuid() != 0:
        sys.exit("Only root can run this script, because the LVM commands need that.\n")

def main(fn):
    dg = DiskGraph(SysInfo())
    print "Graph contains %d entities." % (dg.order - 1, )
    g = dg.todot()
    print "Writing PNG image to %s..." % fn
    g.write_png(fn)
    print "All done!"

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print "Usage: %s <output file>" % sys.argv[0]
        sys.exit(1)
    main(sys.argv[1])



diskgraph/sysinfo.py
# -*- coding: utf-8 -*-
"""Module for collecting information about disks, partitions, RAID arrays and LVM
entities on a Linux server. Part of the diskgraph utility.

Created and tested by Per Rovegård on a server running Ubuntu 11.04 with Python 2.7.1.

Distributed under the 3-Clause BSD license (http://opensource.org/licenses/BSD-3-Clause,
and LICENSE file).
"""

__author__ = "Per Rovegård"
__version__ = "1.2"
__license__ = "BSD-3-Clause"

import subprocess
import re
from check import checker

__all__ = [
    "SysInfo",
    "Root",
    "Partition",
    "LvmPhysicalVolume",
    "LvmVolumeGroup",
    "LvmLogicalVolume",
    "RaidArray",
    "MountedFileSystem",
    "SwapArea",
    "FreeSpace",
]

BLOCK_SIZE = 1024
FREE_SPACE_LIMIT = 100 * 1024 * 1024

def open_file(f):
    with open(f) as fd:
        return [re.split("\\s+", line.strip()) for line in fd if line != ""]

def exec_cmd(args):
    return (re.split("\\s+", line.strip()) for line in subprocess.check_output(args).split("\n") if line != "")

suffixes = ["B", "kB", "MB", "GB", "TB", "PB"]
def tosize(bytesize):
    size = float(bytesize)
    idx = 0
    while size > 1024:
        size /= 1024
        idx += 1
    return "%.2f%s" % (size, suffixes[idx])

class SysObject(object):
    def __str__(self):
        s = "%s\n%s" % (self.gettypename(), self.name)
        if hasattr(self, "byte_size"):
            s += "\n%s" % tosize(self.byte_size)
        return s

    def gettypename(self):
        return self.__class__.__name__

    def expand(self, candidates):
        return [c for c in candidates if c.is_child_of(self)]

    @classmethod
    def generate(cls):
        return []

    def is_child_of(self, tail):
        return False

class Root(SysObject):
    name = "root"

    def __str__(self):
        return "Root"

class FreeSpace(SysObject):
    name = "free"
    def __init__(self, size):
        self.byte_size = size

    @staticmethod
    def is_relevant(size):
        return size >= FREE_SPACE_LIMIT

    def __str__(self):
        return "Free space\n%s" % tosize(self.byte_size)

class Partition(SysObject):
    def __init__(self, line_parts):
        self.kernel_major_minor = (int(line_parts[0]), int(line_parts[1]))
        self.byte_size = int(line_parts[2]) * BLOCK_SIZE
        self.name = line_parts[3]

    def gettypename(self):
        return "Disk" if self.is_disk() else "Partition"

    def is_disk(self):
        return re.match("^[hs]d[a-z]$", self.name)

    def is_partition_for(self, disk):
        return isinstance(disk, Partition) and disk.is_disk() and re.match("^%s\\d+$" % re.escape(disk.name), self.name)

    def is_child_of(self, tail):
        return (isinstance(tail, Root) and self.is_disk()) or self.is_partition_for(tail)

    @classmethod
    def generate(cls):
        return [Partition(p) for p in list(open_file("/proc/partitions"))[2:]]

    def expand(self, candidates):
        result = super(Partition, self).expand(candidates)
        if self.is_disk():
            tot_child_size = sum([c.byte_size for c in result])
            free = self.byte_size - tot_child_size
            if FreeSpace.is_relevant(free):
                result.append(FreeSpace(free))
        return result

class LvmPhysicalVolume(SysObject):
    def __init__(self, parts):
        self.name = parts[0].replace("/dev/", "")
        self.byte_size = int(parts[1])

    def is_child_of(self, tail):
        return isinstance(tail, (Partition, RaidArray)) and tail.name == self.name

    @classmethod
    def generate(cls):
        lines = []
        if checker.has_lvm_commands():
            lines = list(exec_cmd("pvs --noheadings -o pv_name,pv_size --units b --nosuffix".split(" ")))
        return [LvmPhysicalVolume(parts) for parts in lines]

class LvmVolumeGroup(SysObject):
    def __init__(self, parts):
        self.name = parts[0]
        self.byte_size = int(parts[1])
        self.pv_names = [name.replace("/dev/", "") for name in parts[2]]
        self.free_space = int(parts[3])

    def is_child_of(self, tail):
        return isinstance(tail, LvmPhysicalVolume) and tail.name in self.pv_names

    def expand(self, candidates):
        result = super(LvmVolumeGroup, self).expand(candidates)
        if FreeSpace.is_relevant(self.free_space):
            result.append(FreeSpace(self.free_space))
        return result

    @classmethod
    def generate(cls):
        vgs = []
        if checker.has_lvm_commands():
            lines = list(exec_cmd("vgs --noheadings -o vg_name,vg_size,pv_name,vg_free --units b --nosuffix".split(" ")))
            for vg in lines:
                if vgs and vgs[-1][0] == vg[0]:
                    vgs[-1][2].append(vg[2])
                    continue
                vgs.append([vg[0], vg[1], [vg[2]], vg[3]])
        return [LvmVolumeGroup(vg) for vg in vgs]

class LvmLogicalVolume(SysObject):
    def __init__(self, parts):
        self.name = parts[0]
        self.vg_name = parts[1]
        self.byte_size = int(parts[2])

    def is_child_of(self, tail):
        return isinstance(tail, LvmVolumeGroup) and self.vg_name == tail.name

    @classmethod
    def generate(cls):
        lines = []
        if checker.has_lvm_commands():
            lines = list(exec_cmd("lvs --noheadings -o lv_name,vg_name,lv_size --units b --nosuffix".split(" ")))
        return [LvmLogicalVolume(parts) for parts in lines]

class RaidArray(SysObject):
    def __init__(self, data):
        """([name, partition_names...], #blocks)"""
        arr, blocks = data
        self.name = arr[0]
        self.partition_names = arr[1:]
        self.byte_size = blocks * BLOCK_SIZE

    def is_child_of(self, tail):
        return isinstance(tail, Partition) and tail.name in self.partition_names

    @classmethod
    def generate(cls):
        if checker.has_mdstat():
            lines = [line for line in open_file("/proc/mdstat")]
            info = [[i[0]] + [j[:j.find("[")] for j in i[4:] if re.match(".+\\[\\d+\\](\\([A-Z]\\))*", j)] for i in lines if i[0].startswith("md")]
            sizes = [int(line[0]) for line in lines if "blocks" in line]
            return [RaidArray(arr) for arr in zip(info, sizes)]
        return []

class MountedFileSystem(SysObject):
    def __init__(self, parts):
        self.name = parts[5]
        self.path = parts[0]
        self.byte_size = int(parts[1])

    def is_child_of(self, tail):
        if isinstance(tail, (Partition, RaidArray)):
            return "/dev/%s" % tail.name == self.path
        if isinstance(tail, LvmLogicalVolume):
            return "/dev/mapper/%s-%s" % (tail.vg_name, tail.name) == self.path
        return False

    @classmethod
    def generate(cls):
        lines = []
        if checker.has_df_command():
            lines = list(exec_cmd("df -P -B 1".split(" ")))[1:]
        return [MountedFileSystem(parts) for parts in lines]

class SwapArea(SysObject):
    def __init__(self, parts):
        self.name = parts[0].replace("/dev/", "")
        self.byte_size = int(parts[2]) * BLOCK_SIZE

    def is_child_of(self, tail):
        if isinstance(tail, Partition):
            return tail.name == self.name
        return False

    @classmethod
    def generate(cls):
        lines = []
        if checker.has_swaps():
            lines = [line for line in open_file("/proc/swaps")][1:]
        return [SwapArea(parts) for parts in lines]

class SysInfo(object):
    def __init__(self):
        sos = [v for v in globals().values() if isinstance(v, type) and SysObject in v.__bases__]
        self.objects = reduce(lambda x, y: x + y, [so.generate() for so in sos], [])

